doc/developer-guide/generic_type.md

```diff
@@ -0,0 +1,127 @@
+    _x: T;
+
+    constructor(x: T) {
+        this._x = x;
+    }
+    bar<T>(a: T) {
+        console.log('bar: ' + a);
+    }
+}
+```
+
+#### Creation of class type
+
+![](./img/generic_class_type_create.jpg)
+
+The steps to create a generic class type with type arguments are as follows:
+- create class scope  
+  &ensp;The compiler will create a `ClassScope` for each generic class and a `FunctionScope` for each method.
+- create class type  
+  &ensp;Create a generic class type and method types based on `classDeclarationNode`, and associate them with the created `ClassScope` and `FunctionScope`s.
+- create parameters and local variables of method  
+  &ensp;According to `classDeclarationNode`, create parameters and local variables for each method, and save them in the corresponding `FunctionScope`.
+- create the function body of method  
+  &ensp;According to `classDeclarationNode`, create a function body for each method and save it in the corresponding `FunctionScope`.
+
+#### Specialization of class type
+The creation of the specialized class type needs to be based on the `classDeclarationNode` information of the generic class type.
+
+![](./img/generic_class_type_specialize.jpg)
+
+Use `new` to create a new specialized class type:
+- create new class scope  
+  &ensp;Copy the `ClassScope` of the generic class and specialize it.
+- create new class type  
+  &ensp;Copy the type of the generic class and specialize it.
+- specialize parameters and local variables of methods  
+  &ensp;Copy `FunctionScope` of the method in the generic class and specialize the parameters and local variables.
+- specialize the function body of method  
+  &ensp;Specialize function body.
+
+### without type arguments
+``` TypeScript
+class A {
+    _x: number;
+
+    constructor(x: number) {
+        this._x = x;
+    }
+    bar<T>(a: T) {
+        console.log('bar: ' + a);
+    }
+}
+```
+
+#### Creation of class type
+
+![](./img/generic_class_without_typeArgs_type_create.jpg)
+
+The steps to create a generic class type  without type arguments are as follows:
+- create class scope  
+  &ensp;The compiler will create a `ClassScope` for each generic class and a `FunctionScope` for each method.
+- create class type  
+  &ensp;Create a generic class type and method types based on `classDeclarationNode`, and associate them with the created `ClassScope` and `FunctionScope`s.
+- create parameters and local variables of method  
+  &ensp;According to `classDeclarationNode`, create parameters and local variables for each method, and save them in the corresponding `FunctionScope`.
+- create the function body of method  
+  &ensp;According to `classDeclarationNode`, create a function body for each method and save it in the corresponding `FunctionScope`.
+
+#### Specialization of method type
+
+![](./img/generic_class_without_typeArgs_type_specialize.jpg)
+
+When a generic method is called, a specialization of the function is performed:
+- create new method scope  
```

Comment:
So in this situation, we only need to create a new specialized function scope, no need to create a new specialized class scope, right?